share/wake/lib/core/list.wake

```diff
@@ -938,13 +937,44 @@ export def intersect (cmpFn: a => a => Order) (lhs: List a) (rhs: List a): List
 #   subset scmp ("", Nil) ("", "foo", "bar", Nil) = True
 #   subset scmp ("", "bat", Nil) ("", "foo", "bar", Nil) = False
 export def subset (cmpFn: a => a => Order) (lhs: List a) (rhs: List a): Boolean =
-    def lhsSort = sortBy cmpFn lhs
-    def rhsSort = sortBy cmpFn rhs
-
-    match lhsSort rhsSort
+    def subsetImp cmpFn lhsSort rhsSort = match lhsSort rhsSort
         Nil _ -> True
         _ Nil -> False
         (lh, lt) (rh, rt) -> match (cmpFn lh rh)
             LT -> False
-            GT -> subset cmpFn (lh, lt) rt
-            EQ -> subset cmpFn lt rt
+            GT -> subsetImp cmpFn (lh, lt) rt
+            EQ -> subsetImp cmpFn lt rt
+
+    subsetImp cmpFn (sortBy cmpFn lhs) (sortBy cmpFn rhs)
+
+# Returns a list of items where each item was a memeber of lhs but not rhs
+# The returned list will not be the same order as the parameters.
+#
+# Guarantees:
+#  - Output list length will be <= the length of lhs
+#  - All items in output list will be present in lhs
+#  - No items in output list will be present in rhs
+#  - Performance is considered and exceeds the naive N^2 algorithm
+#
+# Examples:
+# ```
+#   difference icmp (seq 4) (seq 3) = 3, Nil
+#   difference icmp (seq 3) (seq 3) = Nil
+#   difference icmp (seq 3) (2, Nil) = 0, 1, Nil
+#   difference scmp ("", "bar", "bat", Nil) ("", "foo", "bat", Nil) = "bar", Nil
```

Comment:
Intersection is symmetric, and "subset" could easily be read as "has as a subset", so I wasn't overly concerned about order.  Here, though, I wonder if we might want to subtract the *left* side from the right -- yes, it's backwards when written traditionally on a single line, but the current order comes across as backwards in pipelines, and it's harder to come up with a reading of the name which would switch the arguments.
```suggestion
#   seq 4 | difference icmp (seq 3) = 3, Nil
#   seq 3 | difference icmp (seq 3) = Nil
#   seq 3 | difference icmp (2, Nil) = 0, 1, Nil
#   ("", "bar", "bat", Nil) | difference scmp ("", "foo", "bat", Nil) = "bar", Nil
```

Reply:
I guess `tsubtract`/`msubtract` have already pinned what order we'd want to use, and it does look like it's `rhs - lhs`.  Let's do follow that model, and maybe bring the names into alignment -- though since List doesn't have a letter prefix and `subtract` doesn't invoke the right mental image, we might want to change to `difference`/`tdifference`/`mdifference` instead, at least if the next release will be a breaking change.

Reply:
I'll just go with `subtract` for now and then we can rename everything to `difference` later. Don't want to mess with even more breaking changes for this release

Reply:
> and it does look like it's rhs - lhs.

Can you double check that? My reading from both is that it is `lhs - rhs`

"Remove all keys from the left which occur in the right"

Reply:
```
sammay@pe01:/scratch/sammay/wake$ wake -x 'tsubtract (listToTree scmp ("a", "b", Nil)) (listToTree scmp ("b", "f", Nil)
) | treeToList'
"a", Nil
sammay@pe01:/scratch/sammay/wake$ wake -x 'msubtract (listToMap scmp ("a" :-> 1, "b" :-> 2, Nil)) (listToMap scmp ("b" :-> 11, "f" :-> 15, Nil)) | mapToList'
Pair "a" 1, Nil
```

My reading of that is "take the left map and remove some keys", and the results support that, but you are definitely right that the wording could be clearer.